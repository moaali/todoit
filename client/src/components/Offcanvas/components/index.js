export  { default as Close } from './Close';
export  { default as NewBody } from './NewBody';
export  { default as TodoBody } from './TodoBody';
