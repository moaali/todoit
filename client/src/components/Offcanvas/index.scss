@import 'settings/index';
@import 'tools/index';
@import 'helpers/index';

.offcanvas {
  // Box Model
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10000;

  // Cosmetics
  background: rgba(0, 0, 0, 0.5);
  visibility: hidden;
  opacity: 0;

  // Ease
  transition: all 100ms ease;

  &.offcanvas--is-open {
    // Cosmetics
    visibility: visible;
    opacity: 1;
  }
}

.offcanvas__body {
  // Box Model
  display: block;
  width: 100%;
  float: right;
  position: relative;
  height: 100%;
  transform: translate3d(100%, 0, 0);

  // Cosmetics
  background: white;
  box-shadow: 0 0 5px 5px rgba(0, 0, 0, 0.06);

  // Ease
  transition: all 180ms ease-in-out;

  @media (min-width: 768px) {
    // Box Model
    width: 600px;
  }

  .offcanvas--is-open & {
    // Box Model
    transform: translate3d(0, 0, 0);
  }
}

.offcanvas__dismiss {
  // Box Model
  display: block;
  width: 60px;
  padding: 15px 0;
  background: white;
  border: 0;
  border-radius: 4px;
  position: absolute;
  top: 0;
  left: calc(50% - 30px);


  @media (min-width: 768px) {
    // Box Model
    top: 60px;
    left: -56px;
  }

  & .fa {
    // Cosmetics
    color: $color-danger;
  }
}
