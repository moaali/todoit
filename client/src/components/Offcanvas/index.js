import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { inject, observer } from 'mobx-react'

import { onKeyPairsPress } from 'services';
import { Close, TodoBody, NewBody } from './components';
import './index.scss'

@inject('store')
@observer
class Offcanvas extends Component {
  static propTypes = {
    store: PropTypes.shape({
      todosStore: PropTypes.shape({}).isRequired,
      offcanvasStore: PropTypes.shape({}).isRequired,
    }).isRequired,
  }

  constructor(props) {
    super(props)
    this.offcanvasStore = props.store.offcanvasStore;
    this.todosStore = props.store.todosStore;
    this.state = {
      todo: {
        title: '',
        description: '',
      },
    };
  }

  componentDidMount() {
    onKeyPairsPress('Escape', undefined, () => {
      if (this.offcanvasStore.isOffcanvasShown)
        this.offcanvasStore.toggleOffcanvasVisibility();
    })
  }

  handleClose = () => {
    const {
      offcanvasStore,
    } = this;

    offcanvasStore.toggleOffcanvasVisibility();
    offcanvasStore.setOffcanvasNewMode(false);
  }

  handleTitleChange = event => {
    const { value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      todo: {
        ...prevState.todo,
        title: value,
      },
    }));
  }

  handleDecriptionChange = event => {
    const { value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      todo: {
        ...prevState.todo,
        description: value,
      },
    }));
  }

  handleSubmit = event => {
    event.preventDefault();

    const {
      state,
      todosStore,
    } = this;

    todosStore.createTodo(state.todo);

    this.handleClose();
  }

  switchViews = todo => {
    const {
      offcanvasStore,
    } = this;

    if (!offcanvasStore.isOffcanvasShown) return null;

    if (offcanvasStore.isOffcanvasNewMode) {
      return (
        <NewBody
          handleTitleChange={this.handleTitleChange}
          handleDecriptionChange={this.handleDecriptionChange}
          handleSubmit={this.handleSubmit}
          title={this.state.todo.title}
          description={this.state.todo.description}
        />
      )
    }

    return <TodoBody todo={todo} />
  }

  render() {
    const todo = this.todosStore.shownTodo;

    return (
      <div className={`offcanvas ${this.offcanvasStore.isOffcanvasShown ? 'offcanvas--is-open' : ''}`}>
        <div className='offcanvas__overlay' />
        <div className='offcanvas__body'>
          <div className="p-60 ov-a h-100p">
            {this.switchViews(todo)}
          </div>
          <Close handleClose={this.handleClose} />
        </div>
      </div>
    );
  }
}

export default Offcanvas;
