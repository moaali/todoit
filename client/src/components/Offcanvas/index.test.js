import React from 'react';
import { Provider } from 'mobx-react';

import stores from 'stores';
import Offcanvas from './';

const OffcanvasComponent = () => (
  <Provider store={{...stores}}>
    <Offcanvas />
  </Provider>
);

describe('Offcanvas', () => {
  it('should match snapshot', () => {
    const component = mount(<OffcanvasComponent />);
    expect(toJson(component)).toMatchSnapshot();
  })

  it('should have `.offcanvas__body` element', () => {
    const component = mount(<OffcanvasComponent />);
    expect(component.find('.offcanvas__body')).toHaveLength(1);
  })
})
