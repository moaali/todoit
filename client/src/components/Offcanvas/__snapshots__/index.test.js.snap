// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offcanvas should match snapshot 1`] = `
<OffcanvasComponent>
  <Provider
    store={
      Object {
        "offcanvasStore": OffcanvasStore {
          "isOffcanvasNewMode": false,
          "isOffcanvasShown": false,
        },
        "searchStore": SearchStore {},
        "sidebarStore": SidebarStore {},
        "todosStore": TodosStore {
          "activeCategory": "all",
          "applyFilters": [Function],
          "currentOffset": 0,
          "displayedTodos": Array [],
          "filterByCategory": [Function],
          "filterByCompleted": [Function],
          "filterByQuery": [Function],
          "filters": Object {
            "category": "all",
            "completed": null,
            "query": "",
          },
          "findById": [Function],
          "loadChunkSize": 10,
          "loadTodos": [Function],
          "meta": Object {
            "all": 0,
            "completed": 0,
            "pending": 0,
          },
          "requestStatus": "error",
          "shownTodo": null,
          "todos": Array [],
        },
      }
    }
  >
    <inject-Offcanvas-with-store>
      <Offcanvas
        store={
          Object {
            "offcanvasStore": OffcanvasStore {
              "isOffcanvasNewMode": false,
              "isOffcanvasShown": false,
            },
            "searchStore": SearchStore {},
            "sidebarStore": SidebarStore {},
            "todosStore": TodosStore {
              "activeCategory": "all",
              "applyFilters": [Function],
              "currentOffset": 0,
              "displayedTodos": Array [],
              "filterByCategory": [Function],
              "filterByCompleted": [Function],
              "filterByQuery": [Function],
              "filters": Object {
                "category": "all",
                "completed": null,
                "query": "",
              },
              "findById": [Function],
              "loadChunkSize": 10,
              "loadTodos": [Function],
              "meta": Object {
                "all": 0,
                "completed": 0,
                "pending": 0,
              },
              "requestStatus": "error",
              "shownTodo": null,
              "todos": Array [],
            },
          }
        }
      >
        <div
          className="offcanvas "
        >
          <div
            className="offcanvas__overlay"
          />
          <div
            className="offcanvas__body"
          >
            <div
              className="p-60 ov-a h-100p"
            />
            <Close
              handleClose={[Function]}
            >
              <button
                className="offcanvas__dismiss"
                onClick={[Function]}
              >
                <i
                  className="fa fa-close"
                />
              </button>
            </Close>
          </div>
        </div>
      </Offcanvas>
    </inject-Offcanvas-with-store>
  </Provider>
</OffcanvasComponent>
`;
